File: android/data/services/remote-gem/src/main/kotlin/com/gemwallet/android/data/services/gemapi/NativeProviderCache.kt
package com.gemwallet.android.data.services.gemapi

import uniffi.gemstone.AlienTarget
import java.security.MessageDigest
import java.util.Base64

internal const val NATIVE_PROVIDER_CACHE_HEADER = "x-gem-cache-ttl"
private const val CACHE_KEY_SEPARATOR: Byte = 0

internal fun AlienTarget.nativeCacheKey(): String? {
    if (headers?.containsKey(NATIVE_PROVIDER_CACHE_HEADER) != true) {
        return null
    }
    val digest = MessageDigest.getInstance("SHA-256")
    digest.update(method.name.toByteArray())
    digest.update(CACHE_KEY_SEPARATOR)
    digest.update(url.toByteArray())
    body?.let {
        digest.update(CACHE_KEY_SEPARATOR)
        digest.update(it)
    }
    return Base64.getEncoder().encodeToString(digest.digest())
}

File: ios/Packages/NativeProviderService/Sources/NativeProviderCache.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import func Gemstone.alienMethodToString
import struct Gemstone.AlienTarget

internal import CryptoKit

let nativeProviderCacheHeader = "x-gem-cache-ttl"
private let cacheKeySeparator = Data([0])

extension AlienTarget {
    var nativeCacheKey: String? {
        guard headers?[nativeProviderCacheHeader] != nil else {
            return nil
        }
        var hasher = SHA256()
        hasher.update(data: Data(alienMethodToString(method: method).utf8))
        hasher.update(data: cacheKeySeparator)
        hasher.update(data: Data(url.utf8))
        if let body {
            hasher.update(data: cacheKeySeparator)
            hasher.update(data: body)
        }
        return Data(hasher.finalize()).base64EncodedString()
    }
}
